src/C++/Parser.cpp

```diff
@@ -85,19 +86,15 @@ EXCEPT ( MessageParseError )
 
       str.assign( m_buffer, 0, pos );
       m_buffer.erase( 0, pos );
-      return true;
+      success = true;
     }
   }
   catch( MessageParseError& e )
   {
-    if( length > 0 )
-      m_buffer.erase( 0, pos + length );
-    else
-      m_buffer.erase();
-
+    m_buffer.erase();
```

Comment:
Length is unable to be greater than zero due to prior executions.  added unit test coverage though to ensure length truly can never be greater than zero at this point even if the length setting method throws an exception